### ISSUE TITLE
Can client cancel observation by returning RST message for non-confirmable notification?

### ISSUE THREAD

**Issue by tamnguyen-tma** [obgm/libcoap#331]
Dear all,

As desciption in RFC7641 about the transmission in CoAP observation (https://tools.ietf.org/pdf/rfc7641.pdf#page=17), client can cancel observer by sending RST message after received unrecognized notification.

```
 If a client rejects a confirmable or non-confirmable notification with a
 Reset message, or if the last attempt to retransmit a confirmable
 notification times out, then the client is considered no longer
 interested and the server MUST remove the associated entry from the
 list of observers.
```

However, I tested with non-confirmable notification, then, there is no any modification in list of observers when client reject the notification. So, is there another way that to cancel observer in this case?

Because the process only handle the notification that added in sendqueue (seems non-confirmable message is not added in this list), therefore, I do not know how server can remove observer out of list when receive the RST message from client.

```
 /* find transaction in sendqueue to stop retransmission */
      coap_remove_from_queue(&context->sendqueue, session, pdu->tid, &sent);

      if (sent) {
        coap_cancel(context, sent);
        if(sent->pdu->type==COAP_MESSAGE_CON && context->nack_handler)
          context->nack_handler(context, sent->session, sent->pdu, COAP_NACK_RST, sent->id);
      }
```

Could you please to help me to solve the problem?
Note: I set type of notification in Get Handler callback function.

Thanks
Tam

**Comment by mrdeep1** [comment]
RFC7641 continues to state immediately following your quote
````
Implementation Note: To properly process a Reset message that
  rejects a non-confirmable notification, a server needs to remember
  the message IDs of the non-confirmable notifications it sends.
  This may be challenging for a server with constrained resources.
  However, since Reset messages are transmitted unreliably, the
  client must be prepared in case the Reset messages are not
  received by the server. Thus, a server can always pretend that a
  Reset message rejecting a non-confirmable notification was lost.
  If a server does this, it could accelerate cancellation by sending
  the following notifications to that client in confirmable
  messages.
````
The current logic does not remember the message IDs for non-confirmable messages, and so would not know what to do with a RST sent from the client (A random RST without anchoring it to the correct message ID could be used to close down information being sent from the server).  To do this enhancement would take some work in the libcoap code.

The alternative way to do this is spelt out in section 3.6.
````
3.6. Cancellation
 ....
 In this case, a client MAY explicitly deregister by issuing a GET
 request that has the Token field set to the token of the observation
 to be cancelled and includes an Observe Option with the value set to
 1 (deregister). All other options MUST be identical to those in the
 registration request except for the set of ETag Options. When the
 server receives such a request, it will remove any matching entry
 from the list of observers and process the GET request as usual.
````
suggests the alternative way of using a value of 1 in the Observe Option which (the last time I looked) does work when sent by the client

**Comment by tamnguyen-tma** [comment]
Hi @mrdeep1 

Thank you for replying me.

>The alternative way to do this is spelt out in section 3.6.

Currently, the libcoap implementation supports only one way to cancel observe for non-confirmable notification, that is Get request with observe Option is set to 1, ins't it? (Can not cancel by RST message).

Tam

**Comment by mrdeep1** [comment]
Correct.  In addition, if the client "goes away" and you are using Coap Ping from the server to the client, the server can detect that the client is "no more" and can delete the coap_session(s) associated with the client which will then delete all the observers associated with that session.

**Comment by tamnguyen-tma** [comment]
Thank you, I understood.

Tam

**Comment by mrdeep1** [comment]
@tamnguyen-tma  See fix in #340

**Comment by tamnguyen-tma** [comment]
Mr. @mrdeep1 

Yes, thank you for supporting this feature.

Tam

**Comment by mrdeep1** [comment]
It should be noted that https://tools.ietf.org/html/rfc8323#section-7.4 states that for CoAP sessions using TCP, the observation must be explicitly cancelled as there is no notion of a RST acknowledgement.

**Comment by mrdeep1** [comment]
@tamnguyen-tma Can this be closed no that #340 is merged into develop.

**Comment by tamnguyen-tma** [comment]
Mr. @mrdeep1 

Yes, sure. There is no problem for this issue.
Thank you.
